### PR TITLE
fix: synchronize properly with tokio::Notify instead of sleeping

### PR DESCRIPTION
**Pull request by mihailjianu1**

### CODE REVIEW

File: rs/bitcoin/adapter/src/router.rs [hunk 1]
```diff
@@ -46,16 +46,15 @@ pub fn start_main_event_loop(
...
         let mut tick_interval = interval(Duration::from_millis(100));
         loop {
-            let sleep_idle_interval = Duration::from_millis(100);
             if adapter_state.is_idle() {
                 connection_manager.make_idle();
                 blockchain_manager.make_idle();
-                // TODO: instead of sleeping here add some async synchronization.
-                sleep(sleep_idle_interval).await;
+                adapter_awake_notifier.notified().await;
                 continue;
```
Comment (rumenov) [inline rs/bitcoin/adapter/src/router.rs:58]: I think you can remove the continue here

File: rs/bitcoin/adapter/src/lib.rs [hunk 2]
```diff
@@ -178,6 +181,12 @@ impl AdapterState {
     pub fn received_now(&self) {
         // Instant::now() is monotonically nondecreasing clock.
         *self.last_received_at.write() = Some(Instant::now());
+        self.awake_notify.notify_waiters();
+    }
+
+    /// Returns a clone of the notify handle.
+    pub fn awake_notifier(&self) -> Arc<Notify> {
```
Comment (rumenov) [inline rs/bitcoin/adapter/src/lib.rs:188]: I would write this as 

```
    pub fn wait_for_request(&self) -> Notified {
        self.awake_notify.notified()
```

and in the other file I would just do

`                adapter_state.wait_for_request().await;
`

**Review by rumenov — APPROVED** [review]

**Review by rumenov — APPROVED** [review]

File: rs/bitcoin/adapter/src/router.rs [hunk 3]
```diff
@@ -50,67 +50,69 @@ pub fn start_main_event_loop(
...
-            let sleep_idle_interval = Duration::from_millis(100);
-            if adapter_state.is_idle() {
-                connection_manager.make_idle();
-                blockchain_manager.make_idle();
-                // TODO: instead of sleeping here add some async synchronization.
-                sleep(sleep_idle_interval).await;
-                continue;
-            }
+            connection_manager.make_idle();
+            blockchain_manager.make_idle();
```
Comment (rumenov) [inline rs/bitcoin/adapter/src/router.rs:54]: nit: I would move the make_idle at the bottom of this loop

File: rs/bitcoin/adapter/src/lib.rs [hunk 4]
```diff
@@ -154,30 +158,64 @@ pub struct AdapterState {
...
 
     /// Updates the current state of the adapter given a request was received.
     pub fn received_now(&self) {
         // Instant::now() is monotonically nondecreasing clock.
         *self.last_received_at.write() = Some(Instant::now());
+        // TODO: perhaps log something if this fails
+        let _ = self.awake_tx.send(());
+    }
+
+    /// A future that returns when/if the adapter becomes/is awake.
```
Comment (rumenov) [inline rs/bitcoin/adapter/src/lib.rs:185]: nit: I think active is a better wording here. Then you can also rename the function to 

`pub fn async active(..);
`

File: rs/bitcoin/adapter/src/router.rs [hunk 5]
```diff
@@ -50,67 +50,69 @@ pub fn start_main_event_loop(
...
+            adapter_state.become_awake().await;
 
             // We do a select over tokio::sync::mpsc::Receiver::recv, tokio::sync::mpsc::UnboundedReceiver::recv,
             // tokio::time::Interval::tick which are all cancellation safe.
```
Comment (rumenov) [inline rs/bitcoin/adapter/src/router.rs:58]: update the comment

Comment (rumenov) [reply rs/bitcoin/adapter/src/router.rs:58]: sorry. nevermind

File: rs/bitcoin/adapter/src/router.rs (hunk 3 above)
Comment (mihailjianu1) [reply rs/bitcoin/adapter/src/router.rs]: Done, makes sense. we don't want to make those idle for nothing, even if we are awake

File: rs/bitcoin/adapter/src/lib.rs [hunk 6]
```diff
@@ -151,33 +159,61 @@ pub struct AdapterState {
...
+        let mut last_received_rx = self.last_received_tx.subscribe();
+        loop {
+            if let Some(last) = *last_received_rx.borrow() {
+                if last.elapsed().as_secs() < self.idle_seconds {
+                    return;
+                }
+            }
+            // Wait for a change in the last received time.
+            let _ = last_received_rx.changed().await;
         }
```
Comment (DSharifi) [inline rs/bitcoin/adapter/src/lib.rs:194]: You can use https://docs.rs/tokio/latest/tokio/sync/watch/struct.Receiver.html#method.wait_for here, and write something like:
```suggestion
        last_received_rx.wait_for(|last| {
          last.is_some_and(|last| {  
             last.elapsed().as_secs()>= self.idle_seconds
          })
        }).await;
```

File: rs/bitcoin/adapter/src/lib.rs (hunk 4 above)
Comment (mihailjianu1) [reply rs/bitcoin/adapter/src/lib.rs:183]: done

File: rs/bitcoin/adapter/src/lib.rs [hunk 7]
```diff
@@ -151,33 +159,61 @@ pub struct AdapterState {
     /// This way the adapter would always be in idle when starting since 'elapsed()' is greater than 'idle_seconds'.
     /// On MacOS this approach caused issues since on MacOS Instant::now() is time since boot and when subtracting
     /// 'idle_seconds' we encountered an underflow and panicked.
-    last_received_at: Arc<RwLock<Option<Instant>>>,
-    /// The field contains how long the adapter should wait to before becoming idle.
-    idle_seconds: u64,
+    ///
+    /// I't simportant that this value is set to None on startup.
```
[the remark this answers is not in the document]
Comment (mihailjianu1) [reply rs/bitcoin/adapter/src/lib.rs]: thanks

File: rs/bitcoin/adapter/src/lib.rs [hunk 8]
```diff
@@ -151,33 +159,61 @@ pub struct AdapterState {
...
+    last_received_tx: watch::Sender<Option<Instant>>,
 }
 
 impl AdapterState {
-    /// Crates new instance of the AdapterState.
+    /// Creates a new instance of the AdapterState.
```
[the remark this answers is not in the document]
Comment (mihailjianu1) [reply rs/bitcoin/adapter/src/lib.rs]: thanks

File: rs/bitcoin/adapter/src/lib.rs (hunk 6 above)
Comment (mihailjianu1) [reply rs/bitcoin/adapter/src/lib.rs]: thanks

File: rs/bitcoin/adapter/src/router.rs (hunk 1 above)
Comment (mihailjianu1) [reply rs/bitcoin/adapter/src/router.rs]: done

**Review by DSharifi — APPROVED** [review]

**Review by rumenov — APPROVED** [review]

**Pull request merged**